lib/pbench/test/unit/server/database/__init__.py

```diff
@@ -0,0 +1,222 @@
+        """
+        Return the first match, or None if there are none left.
+        """
+        return self.selected[0] if self.selected else None
+
+
+class FakeSession:
+    """
+    Mock a SQLAlchemy Session for testing.
+    """
+
+    def __init__(self, cls):
+        """
+        Initialize the context of the session.
+
+        NOTE: 'raise_on_commit' can be set by a caller to cause an exception
+        during the next commit operation. The exception should generally be
+        a subclass of the SQLAlchemy IntegrityError. This is a "one shot" and
+        will be reset when the exception is raised.
+        """
+        self.id = 1
+        self.cls = cls
+        self.added: list[Database.Base] = []
+        self.known: dict[int, Database.Base] = {}
+        self.committed: dict[int, FakeRow] = {}
+        self.rolledback = 0
+        self.queries: list[FakeQuery] = []
+        self.raise_on_commit: Optional[Exception] = None
+
+    def query(self, *entities, **kwargs) -> FakeQuery:
+        """
+        Perform a mocked query on the session, setting up the query context
+        and returning it
+
+        Args:
+            entities: The SQLAlchemy entities on which we're operating
+            kwargs: Additional SQLAlchemy parameters
+
+        Returns:
+            A mocked query object
+        """
+        q = FakeQuery(self)
+        self.queries.append(q)
+        return q
+
+    def add(self, instance: Database.Base):
+        """
+        Add a DB object to a list for testing
+
+        Args:
+            instance: A DB object
+        """
+        self.added.append(instance)
+
+    def commit(self):
+        """
+        Mock a commit operation on the DB session. If the 'raise_on_commit'
+        property has been set, "fail" by raising the exception. Otherwise,
+        mock a commit by updating any cached "committed" values if the "known"
+        proxy objects have changed, and record any new "added" objects.
+        """
+        if self.raise_on_commit:
+            exc = self.raise_on_commit
+            self.raise_on_commit = None
+            raise exc
+        for k, object in self.known.items():
+            self.committed[k] = FakeRow.clone(object)
+        for a in self.added:
+            a.id = self.id
+            self.id += 1
+            for c in a.__table__._columns:
+                if c.default:
+                    default = c.default
+                    # print(f"{type(object).__name__}.{c.name}({type(default).__name__}) default:")
+                    # for x in dir(default):
+                    #     print(f"  {x}: {type(getattr(default, x)).__name__} ({getattr(default, x)})")
+                    if default.is_scalar:
+                        setattr(a, c.name, default.arg)
+                    elif default.is_callable:
+                        setattr(a, c.name, default.arg(None))
+            self.known[a.id] = a
+            self.committed[a.id] = FakeRow.clone(a)
```

Comment:
Should we be concerned about overwriting a "known" item with an "added" item?

If not, would it make sense to add the "added" items to the "known" items first, and then make all the now-known items committed?  (I.e., move the loop at line 185 after the loop at line 187 and omit line 201?)

Reply:
I was thinking that combining the added and known items would make the code simpler and therefore better.

Reply:
The collapsing that I'm suggesting is just in this function (to remove what looks like duplicated code here) -- I didn't mean to suggest any changes to the overall design or implementation.